### PR TITLE
add cc_common.link support for rust_library and rust_test

### PR DESCRIPTION
**Pull request by krasimirgg**
This PR adds support for using cc_common.link to link rust_binary-es and rust_test-s, following https://github.com/bazelbuild/rules_rust/issues/1488:
* a new boolean build setting `--//rules_rust/rust/settings:experimental_use_cc_common_link`, defaulting to false,
* a new rust_toolchain label attribute `experimental_use_cc_common_link`, defaulting to the build setting above.
* a new temporary int `rust_{binary|test|shared_library}.experimental_use_cc_common_link` attribute with possible values `[-1, 0, 1]` defaulting to -1. This is to support gradually moving targets to/from using the new feature; after the feature is stable, we'll remove this attribute.

Given rust `rust_{binary|test}(name = bin)`, and a `rust_toolchain`, `common_cc.link` is enabled if:

* `bin.experimental_use_cc_common_link` is 1, or
* `bin.experimental_use_cc_common_link` is -1 and `rust_toolchain.experimental_use_cc_common_link` points to a label whose value is true.
Otherwise `common_cc.link` is disabled.

When enabled, instead of emitting one `rustc_compile_action` that does both compilation and linking using rustc, we emit 2 actions:
* A rustc one with [–emit=obj](https://doc.rust-lang.org/rustc/command-line-arguments.html#--emit-specifies-the-types-of-output-files-to-generate) to produce an .o file
* A cc_common.link one to produce the final binary (or appropriate binary artefacts for other crates that require linking).

To test this, I've set up a new `test/cc_common_link` directory and [a build bot](https://buildkite.com/bazel/rules-rust-rustlang/builds/6983#018263a7-840d-4894-8a3b-16024ac202f7) to build and test everything there with `---//rules_rust/rust/settings:experimental_use_cc_common_link` set to true. This also acts as an example directory on how to set up a workspace where this is enabled by default.

This feature requires on an appropriate `rust_toolchain.allocator_library`, similarly to how we need one to be able to build a cc_binary with a rust dependency. So I picked up the shared allocator_library repository updates from @bsilver8192's https://github.com/bazelbuild/rules_rust/pull/1350 to set this up.

### CODE REVIEW

File: rust/private/rustc.bzl [hunk 1]
```diff
@@ -1179,6 +1214,52 @@ def rustc_compile_action(
...
+
+        # The path to the package dir, including a trailing "/".
+        package_dir = ctx.bin_dir.path + "/"
+        if ctx.label.workspace_root:
+            package_dir = package_dir + ctx.label.workspace_root + "/"
+        if ctx.label.package:
+            package_dir = package_dir + ctx.label.package + "/"
+
+        if not crate_info.output.path.startswith(package_dir):
+            fail("The package dir path {} should be a suffix of the crate_info.output.path {}", package_dir, crate_info.output.path)
```
Comment (scentini) [inline rust/private/rustc.bzl:1246]:
```suggestion
            fail("The package dir path {} should be a prefix of the crate_info.output.path {}", package_dir, crate_info.output.path)
```

File: rust/private/rustc.bzl [hunk 2]
```diff
@@ -1179,6 +1214,52 @@ def rustc_compile_action(
...
+        linking_contexts = [toolchain.libstd_and_allocator_ccinfo.linking_context, toolchain.stdlib_linkflags.linking_context]
+
+        for dep in crate_info.deps.to_list():
+            if dep.cc_info:
+                linking_contexts.append(dep.cc_info.linking_context)
+
+        # In the cc_common.link action we need to pass the name of the final
+        # binary (output) relative to the package of this target.
+        # We compute it by stripping the path to the package directory,
+        # which is a suffix of the path of `crate_info.output`.
```
Comment (scentini) [inline rust/private/rustc.bzl:1236]:
```suggestion
        # which is a prefix of the path of `crate_info.output`.
```

**Review by scentini — APPROVED** [review]

**Comment by bsilver8192** [comment]
Looking at the CI task added here (as part of adding another one), it looks like it runs the tests without the flag enabled. Is that intentional?

**Comment by bsilver8192** [comment]
From looking at the rest of that file, I'm pretty sure you have to duplicate it under `test_flags` to also apply it to the tests.

**Pull request merged**